src/client/websocket/handlers/TYPING_START.js

```diff
@@ -5,14 +5,50 @@ const { Events } = require('../../../util/Constants');
 module.exports = (client, { d: data }) => {
   const channel = client.channels.cache.get(data.channel_id);
   const user = client.users.cache.get(data.user_id);
+  const timestamp = new Date(data.timestamp * 1000);
+  let timeout;
 
   if (channel && user) {
-    /**
-     * Emitted whenever a user starts typing in a channel.
-     * @event Client#typingStart
-     * @param {Channel} channel The channel the user started typing in
-     * @param {User} user The user that started typing
-     */
-    client.emit(Events.TYPING_START, channel, user);
+    if (channel.type === 'voice') {
+      client.emit(Events.WARN, `Discord sent a typing packet to a voice channel ${channel.id}`);
+      return;
+    }
+
+    if (channel._typing.has(user.id)) {
+      const typing = channel._typing.get(user.id);
+
+      typing.lastTimestamp = timestamp;
+      typing.elapsedTime = getElapsedTime(typing.lastTimestamp, typing.since);
+      resetTimeout(client, tooLate(channel, user), timeout);
+    } else {
+      let since = new Date();
+      let lastTimestamp = new Date();
+      resetTimeout(client, tooLate(channel, user), timeout);
+      channel._typing.set(user.id, { user, since, lastTimestamp, elapsedTime: getElapsedTime(lastTimestamp, since) });
+
+      /**
+       * Emitted whenever a user starts typing in a channel.
+       * @event Client#typingStart
+       * @param {Channel} channel The channel the user started typing in
+       * @param {User} user The user that started typing
+       */
+      client.emit(Events.TYPING_START, channel, user);
+    }
   }
 };
+
+function getElapsedTime(since) {
+  return Date.now() - since;
+}
+
+function resetTimeout(client, _timeout, timeout) {
+  client.clearTimeout(timeout);
+  timeout = _timeout;
```

Comment:
this redefines the parameter, not the variable you defined in the handler function :man_facepalming:

Reply:
Woops, I overlooked that, my bad! Thanks for pointing it out